govcd/ui_plugin_test.go

```diff
@@ -0,0 +1,269 @@
+//go:build functional || openapi || uiPlugin || ALL
+
+package govcd
+
+import (
+	"fmt"
+	"net/http"
+	"net/textproto"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	. "gopkg.in/check.v1"
+)
+
+func init() {
+	testingTags["uiPlugin"] = "ui_plugin_test.go"
+}
+
+// This object is equivalent to the manifest.json that is inside the ../test-resources/ui_plugin.zip file
+var testUIPluginMetadata = &types.UIPluginMetadata{
+	Vendor:         "VMware",
+	License:        "BSD-2-Clause",
+	Link:           "http://www.vmware.com",
+	PluginName:     "Test Plugin",
+	Version:        "1.2.3",
+	Description:    "Test Plugin description",
+	ProviderScoped: true,
+	TenantScoped:   true,
+}
+
+// Test_UIPlugin tests all the possible operations that can be done with a UIPlugin object in VCD.
+func (vcd *TestVCD) Test_UIPlugin(check *C) {
+	if vcd.skipAdminTests {
+		check.Skip(fmt.Sprintf(TestRequiresSysAdminPrivileges, check.TestName()))
+	}
+
+	// Add a plugin present on disk
+	newUIPlugin, err := vcd.client.AddUIPlugin("../test-resources/ui_plugin.zip", true)
+	check.Assert(err, IsNil)
+	AddToCleanupListOpenApi(newUIPlugin.UIPluginMetadata.ID, check.TestName(), types.OpenApiEndpointExtensionsUi+newUIPlugin.UIPluginMetadata.ID)
```

Comment:
I think it is worth adding a note that due to a bug the cleanup will return an error if there is nothing to cleanup so that people don't get surprised
```
# 2 of 2 - removeLeftoverEntries: [INFO] Attempting cleanup of OpenApiEntity 'urn:vcloud:uiPlugin:b23f0760-1c2a-4ffd-a43a-8531779205fe' instantiated by TestVCD.Test_UIPlugin
removeLeftoverEntries: [ERROR] Error deleting OpenApiEntity 'urn:vcloud:uiPlugin:b23f0760-1c2a-4ffd-a43a-8531779205fe': error in HTTP GET request: INTERNAL_SERVER_ERROR - [ 2dc92c82-fff5-47dd-b950-d4cd458fd530 ] class java.lang.NullPointerException
```

Reply:
Actually @dataclouder hit precisely this one 🎯 
I have added this snippet to the `removeLeftoverEntities` function:
```
		if strings.Contains(entity.OpenApiEndpoint, types.OpenApiEndpointExtensionsUi) {
			err = amendUIPluginGetByIdError(entity.Name, err)
		}
```